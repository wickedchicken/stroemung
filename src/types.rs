use crate::math::Real;
use ndarray::{Array, Ix2};

pub type Velocity = [Real; 2];

pub type GridArray<T> = Array<T, Ix2>;

pub type GridSize = [usize; 2];

// According to https://docs.rs/ndarray/0.16.1/ndarray/struct.ArrayBase.html,
// "The default memory order of an array is row major order (a.k.a “c” order),
// where each row is contiguous in memory." This means we can use the default
// implementation of PartialOrd and Ord. If we were using column-major ordering,
// we would need to implement an ordering that prioritized the y value over the
// x value.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct BoundaryIndex(pub usize, pub usize);

// It would be nice to unify BoundaryIndex and GridIndex into one type that
// can be sorted and also directly used by ndarray's indexing operations.
pub type GridIndex = (usize, usize);
