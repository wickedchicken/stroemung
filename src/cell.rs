use serde::{Deserialize, Serialize};

use crate::types::Velocity;
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum BoundaryCell {
    Inflow { velocity: Velocity },
    Outflow,
    NoSlip,
}

impl fmt::Display for BoundaryCell {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum Cell {
    Fluid,
    Boundary(BoundaryCell),
}

impl fmt::Display for Cell {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}
