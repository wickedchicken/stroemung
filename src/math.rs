use ndarray::ArrayView2;

pub type Real = f64;

/// Calculate du^2/dx (the derivative of u^2 over x)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the row where j is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `delx` - "delta x," the physical width of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn du2dx(u_view: ArrayView2<Real>, delx: Real, gamma: Real) -> Real {
    let u_i_m1 = u_view[(0, 1)]; // u[(i-1, j)]  "u[i minus 1]" -> u_i_m1
    let u_i = u_view[(1, 1)]; // u[(i, j)]  "u[i]" -> "u_i"
    let u_i_p1 = u_view[(2, 1)]; // u[(i+1, j)]  "u[i plus 1]" -> u_i_p1

    let inner_left1 = (u_i + u_i_p1).powi(2);
    let inner_right1 = (u_i_m1 + u_i).powi(2);

    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (u_i + u_i_p1).abs() * (u_i - u_i_p1);
    let inner_right2 = (u_i_m1 + u_i).abs() * (u_i_m1 - u_i);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * delx)
}

/// Calculate duv/dx (the derivative of u*v over x)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses the four
///   "lower left" values (the four combinarions of i-1, i, j, and j+1), but
///   takes a 3x3 ArrayView2 to be easier to combine with other functions.
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the row where j is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `delx` - "delta x," the physical width of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn duvdx(
    u_view: ArrayView2<Real>,
    v_view: ArrayView2<Real>,
    delx: Real,
    gamma: Real,
) -> Real {
    let u_i_j = u_view[(1, 1)]; // u[(i, j)] -> u_i_j
    let u_i_j_p1 = u_view[(1, 2)]; // u[(i, j+1)]  "u[i][j plus 1]" -> u_i_j_p1
    let u_i_m1_j = u_view[(0, 1)]; // "u[(i-1, j)]" "u[i minus 1][j]" -> u_i_m1_j
    let u_i_m1_j_p1 = u_view[(0, 2)]; // "u[(i-1, j+1)]" "u[i-1][j+1]" -> u_i_m1_j_p1

    let v_i_j = v_view[(1, 1)]; // "v[(i, j)]" -> v_i_j
    let v_i_p1_j = v_view[(2, 1)]; // "v[(i+1, j)]" -> "v[i plus 1][j]" -> v_i_p1_j
    let v_i_m1_j = v_view[(0, 1)]; // "v[(i-1, j)]" -> "v[i-1][j]" -> v_i_m1_j

    let inner_left1 = (u_i_j + u_i_j_p1) * (v_i_j + v_i_p1_j);
    let inner_right1 = (u_i_m1_j + u_i_m1_j_p1) * (v_i_m1_j + v_i_j);

    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (u_i_j + u_i_j_p1).abs() * (v_i_j - v_i_p1_j);
    let inner_right2 = (u_i_m1_j + u_i_m1_j_p1).abs() * (v_i_m1_j - v_i_j);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * delx)
}

/// Calculate duv/dy (the derivative of u*v over y)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `u_view` - A 3x3-element ArrayView2 representing
///   u[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the column where i is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   "upper right" values (the four combinarions of i, i+1, j-1, and j), but
///   takes a 3x3 ArrayView2 to be easier to combine with other functions.
/// * `dely` - "delta y," the physical height of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn duvdy(
    u_view: ArrayView2<Real>,
    v_view: ArrayView2<Real>,
    dely: Real,
    gamma: Real,
) -> Real {
    let u_i_j = u_view[(1, 1)]; // u[(i, j)] -> u_i_j
    let u_i_j_m1 = u_view[(1, 0)]; // u[(i, j-1)] -> "u[i][j minus 1]" -> u_i_j_m1
    let u_i_j_p1 = u_view[(1, 2)]; // u[(i, j+1)] -> "u[i][j plus 1]" -> u_i_j_p1

    let v_i_j = v_view[(1, 1)]; // v[(i, j)] -> v_i_j
    let v_i_j_m1 = v_view[(1, 0)]; // v[(i, j-1)] -> "v[i][j minus 1]" -> v_i_j_m1
    let v_i_p1_j = v_view[(2, 1)]; // v[(i+1, j)] -> "v[i plus 1][j]" -> v_i_p1_j
    let v_i_p1_j_m1 = v_view[(2, 0)]; // v[(i+1, j-1)] -> "v[i plus 1][j minus 1]" -> v_i_p1_j_m1

    let inner_left1 = (v_i_j + v_i_p1_j) * (u_i_j + u_i_j_p1);
    let inner_right1 = (v_i_j_m1 + v_i_p1_j_m1) * (u_i_j_m1 + u_i_j);
    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (v_i_j + v_i_p1_j).abs() * (u_i_j - u_i_j_p1);
    let inner_right2 = (v_i_j_m1 + v_i_p1_j_m1).abs() * (u_i_j_m1 - u_i_j);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * dely)
}

/// Calculate dv^2/dy (the derivative of v^2 over y)
///
/// This function uses the same basic algebra rearrangement that the
/// NaSt2D code does. This makes the function easier to compare against
/// NaSt2D (and reduces the number of divisions in the calculation).
///
/// # Arguments
///
/// * `v_view` - A 3x3-element ArrayView2 representing
///   v[(i-1) to (i+1), (j-1) to (j+1)]. This function only uses
///   the three values on the column where i is 0 (index 1), but takes a 3x3
///   ArrayView2 to be easier to combine with other functions.
/// * `dely` - "delta y," the physical width of the cell
/// * `gamma` - Greek letter gamma, the upwind discretization parameter
pub fn dv2dy(v_view: ArrayView2<Real>, dely: Real, gamma: Real) -> Real {
    let v_i_j = v_view[(1, 1)]; // v[(i, j)] -> v_i_j
    let v_i_j_p1 = v_view[(1, 2)]; // v[(i, j+1)] -> "v[i][j plus 1]" -> v_i_j_p1
    let v_i_j_m1 = v_view[(1, 0)]; // v[(i, j-1)] -> "v[i][j minus 1]" -> v_i_j_m1

    let inner_left1 = (v_i_j + v_i_j_p1).powi(2);
    let inner_right1 = (v_i_j_m1 + v_i_j).powi(2);

    let left_side = inner_left1 - inner_right1;

    let inner_left2 = (v_i_j + v_i_j_p1).abs() * (v_i_j - v_i_j_p1);
    let inner_right2 = (v_i_j_m1 + v_i_j).abs() * (v_i_j_m1 - v_i_j);

    (left_side + (gamma * (inner_left2 - inner_right2))) / (4.0 * dely)
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::{array, ArrayView2};

    #[test]
    fn test_du2dx() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[0., 1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.,
                1.7,
                3.15,
            ),
            (
                array![[0., -1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.,
                1.7,
                5.15,
            ),
            (
                array![[0., 1., 0.], [0., 1., 0.], [0., 1., 0.]],
                1.,
                1.7,
                0.,
            ),
            (
                array![[0., 1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.,
                1.3,
                3.35,
            ),
            (
                array![[0., 1., 0.], [0., 2., 0.], [0., 3., 0.]],
                1.5,
                1.7,
                2.1,
            ),
            (
                array![[0., 10., 0.], [0., 20., 0.], [0., 30., 0.]],
                1.5,
                1.7,
                210.,
            ),
        ];
        for (u, delx, gamma, expected) in test_cases {
            assert_eq!(du2dx(ArrayView2::from(&u), delx, gamma), expected);
        }
    }

    #[test]
    fn test_duvdx() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.7,
                40.35,
            ),
            (
                array![[1., 2., 3.], [4., 5., -6.], [-7., 8., 9.]],
                array![[8., 9., 10.], [11., -12., 13.], [14., 15., -16.]],
                1.,
                1.7,
                -53.1,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.6,
                1.7,
                25.218750,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.5,
                41.25,
            ),
        ];
        for (u, v, delx, gamma, expected) in test_cases {
            assert_eq!(
                duvdx(ArrayView2::from(&u), ArrayView2::from(&v), delx, gamma),
                expected
            );
        }
    }

    #[test]
    fn test_duvdy() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.7,
                17.15,
            ),
            (
                array![[1., 2., 3.], [4., 5., -6.], [-7., 8., 9.]],
                array![[8., 9., 10.], [11., -12., 13.], [14., 15., -16.]],
                1.,
                1.7,
                -32.35,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.6,
                1.7,
                10.718749999999998,
            ),
            (
                array![[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]],
                array![[8., 9., 10.], [11., 12., 13.], [14., 15., 16.]],
                1.,
                1.5,
                17.25,
            ),
        ];
        for (u, v, dely, gamma, expected) in test_cases {
            assert_eq!(
                duvdy(ArrayView2::from(&u), ArrayView2::from(&v), dely, gamma),
                expected
            );
        }
    }

    #[test]
    fn test_dv2dy() {
        // These don't have any particular significance, just some random data.
        let test_cases = [
            (
                array![[0., 0., 0.], [1., 2., 3.], [0., 0., 0.]],
                1.,
                1.7,
                3.15,
            ),
            (
                array![[0., 0., 0.], [-1., 2., 3.], [0., 0., 0.]],
                1.,
                1.7,
                5.15,
            ),
            (
                array![[0., 0., 0.], [1., 1., 1.], [0., 0., 0.]],
                1.,
                1.7,
                0.,
            ),
            (
                array![[0., 0., 0.], [1., 2., 3.], [0., 0., 0.]],
                1.,
                1.3,
                3.35,
            ),
            (
                array![[0., 0., 0.], [1., 2., 3.], [0., 0., 0.]],
                1.5,
                1.7,
                2.1,
            ),
            (
                array![[0., 0., 0.], [10., 20., 30.], [0., 0., 0.]],
                1.5,
                1.7,
                210.,
            ),
        ];
        for (v, dely, gamma, expected) in test_cases {
            assert_eq!(dv2dy(ArrayView2::from(&v), dely, gamma), expected);
        }
    }
}
