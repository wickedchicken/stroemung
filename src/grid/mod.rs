pub mod presets;

use std::collections::BTreeSet;
use std::fmt;
use std::io::Read;

use serde::Deserialize;
use serde::Serialize;

use serde_json::Error as SerdeError;

use ndarray::Zip;
use thiserror::Error;

use crate::cell::Cell;
use crate::math::Real;
use crate::types::{BoundaryIndex, GridArray, GridIndex, GridSize};

type Neighbors = [Option<GridIndex>; 4];

#[derive(Error, Debug)]
pub enum SimulationGridError {
    #[error("An error occurred while deserializing: `{0}`")]
    DeserializationError(#[from] SerdeError),
    #[error("A cell `{0}` at `{1}` was not a BoundaryCell as expected.")]
    BoundaryListIncorrectError(String, String),
    #[error("A cell `{0}` at `{1}` has fluid on opposing sides.")]
    BoundaryTooThinError(String, String),
}

#[derive(Debug, Default)]
pub struct BoundaryList {
    boundaries: BTreeSet<BoundaryIndex>,
    pub sorted_boundary_list: Vec<(GridIndex, Neighbors)>,
}

impl std::fmt::Display for BoundaryList {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Boundaries:")?;
        for elem in &self.boundaries {
            writeln!(f, "  {:?}", elem)?;
        }
        writeln!(f, "Sorted Boundary List:")?;
        for elem in &self.sorted_boundary_list {
            writeln!(f, "  {:?}", elem)?;
        }
        Ok(())
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UnfinalizedSimulationGrid {
    size: GridSize,
    pressure: GridArray<Real>,
    u: GridArray<Real>,
    v: GridArray<Real>,
    cell_type: GridArray<Cell>,
}

// Useful for test code
impl From<SimulationGrid> for UnfinalizedSimulationGrid {
    fn from(item: SimulationGrid) -> Self {
        // Will be nicer once https://github.com/rust-lang/rust/issues/86555
        // is in stable.
        UnfinalizedSimulationGrid {
            size: item.size,
            pressure: item.pressure,
            u: item.u,
            v: item.v,
            cell_type: item.cell_type,
        }
    }
}

// This must be the same as UnfinalizedSimulationGrid, except for boundaries.
// We have two types to make sure we never deserialize without forgetting to
// generate the boundary list.
#[derive(Debug, Serialize)]
pub struct SimulationGrid {
    pub size: GridSize,
    pub pressure: GridArray<Real>,
    pub u: GridArray<Real>,
    pub v: GridArray<Real>,
    pub cell_type: GridArray<Cell>,
    #[serde(skip)]
    pub boundaries: BoundaryList,
}

impl TryFrom<UnfinalizedSimulationGrid> for SimulationGrid {
    type Error = SimulationGridError;

    fn try_from(item: UnfinalizedSimulationGrid) -> Result<Self, Self::Error> {
        // Will be nicer once https://github.com/rust-lang/rust/issues/86555
        // is in stable.
        let mut grid = SimulationGrid {
            size: item.size,
            pressure: item.pressure,
            u: item.u,
            v: item.v,
            cell_type: item.cell_type,
            boundaries: Default::default(),
        };
        grid.rebuild_boundary_list()?;
        Ok(grid)
    }
}

impl std::fmt::Display for SimulationGrid {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Simulation grid {}x{}", self.size[0], self.size[1],)?;
        writeln!(f, "Pressure:{}", self.pressure)?;
        writeln!(f, "u:{}", self.u)?;
        writeln!(f, "v:{}", self.v)?;
        writeln!(f, "Cell Type:{}", self.cell_type)?;
        Ok(())
    }
}

impl SimulationGrid {
    fn rebuild_boundary_list(&mut self) -> Result<(), SimulationGridError> {
        self.boundaries.boundaries.clear();
        // Run a for_each with the value and indices. See
        // https://github.com/rust-ndarray/ndarray/issues/1093 for details.
        Zip::indexed(self.cell_type.view()).for_each(|idx, val| {
            if let Cell::Boundary(_) = val {
                self.boundaries
                    .boundaries
                    .insert(BoundaryIndex(idx.0, idx.1));
            }
        });
        self.boundaries.sorted_boundary_list = self
            .boundaries
            .boundaries
            .iter()
            .copied()
            .map(|x| ((x.0, x.1), self.fluid_neighbors((x.0, x.1))))
            .collect();

        for (idx, [left, right, up, down]) in &self.boundaries.sorted_boundary_list {
            if (left.is_some() && right.is_some()) || (up.is_some() && down.is_some()) {
                return Err(SimulationGridError::BoundaryTooThinError(
                    self.cell_type[*idx].to_string(),
                    format!("{:?}", *idx),
                ));
            }
        }
        Ok(())
    }

    fn fluid_neighbors(&self, cell_idx: GridIndex) -> Neighbors {
        let left: Option<GridIndex> = if cell_idx.0 > 0 {
            let test_index = (cell_idx.0 - 1, cell_idx.1);
            match self.cell_type[test_index] {
                Cell::Fluid => Some(test_index),
                _ => None,
            }
        } else {
            None
        };

        let right: Option<GridIndex> = if cell_idx.0 < (self.size[0] - 1) {
            let test_index = (cell_idx.0 + 1, cell_idx.1);
            match self.cell_type[test_index] {
                Cell::Fluid => Some(test_index),
                _ => None,
            }
        } else {
            None
        };

        let up: Option<GridIndex> = if cell_idx.1 > 0 {
            let test_index = (cell_idx.0, cell_idx.1 - 1);
            match self.cell_type[test_index] {
                Cell::Fluid => Some(test_index),
                _ => None,
            }
        } else {
            None
        };

        let down: Option<GridIndex> = if cell_idx.1 < (self.size[1] - 1) {
            let test_index = (cell_idx.0, cell_idx.1 + 1);
            match self.cell_type[test_index] {
                Cell::Fluid => Some(test_index),
                _ => None,
            }
        } else {
            None
        };

        [left, right, up, down]
    }

    pub fn from_reader<R: Read>(
        reader: R,
    ) -> Result<SimulationGrid, SimulationGridError> {
        match serde_json::from_reader::<R, UnfinalizedSimulationGrid>(reader) {
            Ok(unfinalized) => SimulationGrid::try_from(unfinalized),
            Err(x) => Err(SimulationGridError::DeserializationError(x)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array;
    use std::fs::File;
    use std::io::BufReader;
    use std::path::{Path, PathBuf};

    fn test_data_directory() -> PathBuf {
        Path::new(file!())
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .join("test_data")
    }

    #[test]
    fn grid_size() {
        let size = [5, 7];
        let grid = presets::empty(size);
        assert_eq!(grid.size[0], 5);
        assert_eq!(grid.size[1], 7);
        assert_eq!(grid.pressure.shape(), size);
        assert_eq!(grid.u.shape(), size);
        assert_eq!(grid.v.shape(), size);
        assert_eq!(grid.cell_type.shape(), size);
    }

    #[test]
    fn thin_boundary() {
        use crate::cell::{BoundaryCell, Cell};
        let size = [3, 3];

        let boundaries: Vec<Vec<GridIndex>> =
            vec![vec![(1, 0), (1, 1), (1, 2)], vec![(0, 1), (1, 1), (2, 1)]];
        for example in &boundaries {
            let mut unfinalized = UnfinalizedSimulationGrid {
                size,
                pressure: Array::zeros(size),
                u: Array::zeros(size),
                v: Array::zeros(size),
                cell_type: Array::from_elem(size, Cell::Fluid),
            };
            for idx in example {
                unfinalized.cell_type[*idx] = Cell::Boundary(BoundaryCell::NoSlip);
            }
            let grid = SimulationGrid::try_from(unfinalized);
            assert!(grid.is_err());
            assert!(format!("{:?}", grid).contains("BoundaryTooThinError"));
        }
    }

    #[test]
    fn rebuild_boundary_list() {
        use crate::cell::{BoundaryCell, Cell};
        let size = [3, 3];
        let mut unfinalized = UnfinalizedSimulationGrid {
            size,
            pressure: Array::zeros(size),
            u: Array::zeros(size),
            v: Array::zeros(size),
            cell_type: Array::from_elem(size, Cell::Fluid),
        };

        // Everything except for the middle cell
        let expected_boundaries: Vec<GridIndex> = vec![
            (0, 0),
            (0, 1),
            (0, 2),
            (1, 0),
            (1, 2),
            (2, 0),
            (2, 1),
            (2, 2),
        ];
        let expected_boundary_indices: Vec<BoundaryIndex> = expected_boundaries
            .iter()
            .map(|x| BoundaryIndex(x.0, x.1))
            .collect();

        let expected_neighbors: Vec<Neighbors> = vec![
            [None, None, None, None],
            [None, Some((1, 1)), None, None],
            [None, None, None, None],
            [None, None, None, Some((1, 1))],
            [None, None, Some((1, 1)), None],
            [None, None, None, None],
            [Some((1, 1)), None, None, None],
            [None, None, None, None],
        ];

        let expected_sorted_list: Vec<(GridIndex, Neighbors)> = expected_boundaries
            .iter()
            .zip(expected_neighbors)
            .map(|(x, y)| (*x, y))
            .collect();

        for idx in &expected_boundaries {
            unfinalized.cell_type[*idx] = Cell::Boundary(BoundaryCell::NoSlip);
        }

        let grid = SimulationGrid::try_from(unfinalized).unwrap();

        let calculated_boundaries_as_list: Vec<BoundaryIndex> =
            grid.boundaries.boundaries.iter().copied().collect();

        assert_eq!(calculated_boundaries_as_list, expected_boundary_indices);
        assert_eq!(grid.boundaries.sorted_boundary_list, expected_sorted_list);
    }

    #[test]
    fn deserialize() {
        let test_filename = test_data_directory().join("simple_grid.json");
        let result = SimulationGrid::from_reader(BufReader::new(
            File::open(test_filename).unwrap(),
        ))
        .unwrap();
        insta::assert_json_snapshot!(result);
    }

    #[test]
    fn deserialize_boundaries() {
        let test_filename = test_data_directory().join("small_grid_with_boundaries.json");
        let result = SimulationGrid::from_reader(BufReader::new(
            File::open(test_filename).unwrap(),
        ))
        .unwrap();

        insta::assert_json_snapshot!(result);
        insta::assert_snapshot!(result.boundaries);
    }

    #[test]
    fn serialize() {
        let size = [2, 3];
        let grid = presets::empty(size);
        insta::assert_json_snapshot!(grid);
    }
}
