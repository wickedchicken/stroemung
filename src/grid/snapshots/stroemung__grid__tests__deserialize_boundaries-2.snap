---
source: src/grid/mod.rs
expression: result.boundaries
---
Boundaries:
  BoundaryIndex(0, 0)
  BoundaryIndex(0, 1)
  BoundaryIndex(0, 2)
  BoundaryIndex(1, 0)
  BoundaryIndex(1, 2)
  BoundaryIndex(2, 0)
  BoundaryIndex(2, 2)
  BoundaryIndex(3, 0)
  BoundaryIndex(3, 1)
  BoundaryIndex(3, 2)
Sorted Boundary List:
  ((0, 0), [None, None, None, None])
  ((0, 1), [None, Some((1, 1)), None, None])
  ((0, 2), [None, None, None, None])
  ((1, 0), [None, None, None, Some((1, 1))])
  ((1, 2), [None, None, Some((1, 1)), None])
  ((2, 0), [None, None, None, Some((2, 1))])
  ((2, 2), [None, None, Some((2, 1)), None])
  ((3, 0), [None, None, None, None])
  ((3, 1), [Some((2, 1)), None, None, None])
  ((3, 2), [None, None, None, None])
